src/webapp/cilogon_ldap.py

```diff
@@ -42,3 +42,33 @@ def cilogon_id_map_to_ssh_keys(m):
         if 'sshPublicKey' in v['data']
     }
 
+
+def _entry2cinfo(entry):
+    ci = {'DNs': [entry['dn']]}
+    emails = entry['data'].get('mail')
+    if emails:
+        ci['PrimaryEmail'] = emails[0].lower()
+        if len(emails) >= 2:
+            ci['SecondaryEmail'] = emails[1].lower()
+    else:
+        return None
+    return ci
+
+
+def cilogon_id_map_to_yaml_data(m):
+    return {
+        id_: {'CILogonID'          : id_,
+              'FullName'           : entry['data']['cn'],
+              'ContactInformation' : cinfo}
+        for id_, entry in m.items()
+        for cinfo in [_entry2cinfo(entry)] if cinfo
```

Comment:
wait so is `[_entry2cinfo(entry)]` always going to be a 1-element list, with the element being either a dict or a None?

Reply:
Yeah, exactly.  The point of 392fed4 was to exclude the whole contact if the entry lacks an email address.  It seemed like the most straightforward way to add that filtering to the dict comprehension (without a redundant `_entry2cinfo(entry)` call).  But feel free to suggest if you have something you like better in mind.

Reply:
If you have to pull trickery like that in order to get the dict comprehension to work, then it's time to stop using a dict comprehension and do it the old fashioned way.
```
    ret = {}
    for id_, entry in m.items():
        cinfo = _entry2cinfo(entry)
        if not cinfo:
            continue
        ret[id_] = {'CILogonID'          : id,
                    'FullName'           : entry['data']['cn'],
                    'ContactInformation' : cinfo}
    return ret
```

Reply:
Eh sure if that's how you'd like it.

For me, adding an extra `for var in [some_expression()]` line is kind of my standard idiom to assign an intermediate value to a var in a list/dict comprehension / generator expression.  It just feels more pythonic to me than building up an object in a regular for-loop.

(As a piece of trivial knowledge, apparently list comprehensions have been part of python since the year 2000.  I can't even comprehend how they did the special effects in The Matrix without that feature.)

Reply:
Updated to the Old Fashioned Burger